### PR TITLE
Containerized Playbook Runs

### PR DESCRIPTION
**Pull request by jladdjr**
Branching off of https://github.com/ansible/ansible-runner/pull/446

Depends-On: https://github.com/ansible/project-config/pull/533

### CODE REVIEW

**Comment by jladdjr** [comment]
Copying over @kdelee's testing checklist from https://github.com/ansible/ansible-runner/pull/446#issuecomment-649684538:

Changes requested:
---------------------------

- [x] If user has `containerized: true` but not able to launch with `container_runtime`, fail gracefully. 
- [ ] ~.. and have `containerized` internally default to `false`~ (jladd: not sure that if user requests containerized execution that we want to switch to a regular run if we don't find the container engine)
- [ ] ~If user does not set `container_runtime` try to use first `podmand` then `docker` if neither launch, fail gracefully~ (update: jladd: I'm a little iffy on this. think that we want to avoid surprising a user by switching out the container engines)
- [x] `container_name` need setting for settings file for container name which is assumed to include tag and/or registry name (FQContainerName). Default to `ansible/ansible-runner`
- [ ] ????? `container_pull` default true because this is what docker will do for us. For some use cases people may want to make sure not to pull before launching, using whatever is available or failing if not locally available. Not sure if we want to deal w/ this at the runner level or not.
- [x] `container_volume_mounts` is a settings file option but seems to have some problems around whitspace. Confirm it can take multiple volumes to mount. Additionally right now only supports "symmetric" mounts, e.g. `"/tmp:/tmp" but is common docker use case to mount different local dir to dir in container. As a minimum need better documention of option and whitespace handling
- [x] add cli flag for `containerized`
- [x] add cli flag for `container-runtime`
- [x] add cli flag for ~`container-name`~ `container-image`
- [x] add cli flag for `container-volume-mounts`
- [ ] ???? if add `container_pull` option then add `container-pull` cli option
~- [ ] add env var for `RUNNER_CONTAINERIZED`~
~- [ ] add env var for `RUNNER_CONTAINER_RUNTIME`~
~- [ ] add env var for `RUNNER_CONTAINER_NAME`~
~- [ ] add env var for `RUNNER_CONTAINER_VOLUME_MOUNTS`~
~- [ ] add env var for `RUNNER_CONTAINER_PORTS`~
- [ ] ???? if add `container_pull` option then `RUNNER_CONTAINER_PULL`
- [x] ensure all internal methods accept all container options
- [x] if container does not have whatever we need to call ansible-playbook, handle error gracefully and place something in the artifacts file and not just error to stdout so that when user that consumes artifacts to find results, they have evidence that helps them find out what happened.



Test case ideas:
----------------------
Accept settings + Run Job + Artifacts are present:
  - configured w/ settings file ✅ 
  - ~configured w/ env vars~
  - configured w/ cli options ✅ 
  - use a playbook that looks at facts that can tell us if in a container, and that fails if it is not in a container ✅ 
  - check that job result in the tests ✅ 

Throws error if `containerized` is true but can't use container runtime (use foobar):
  - use easiest setting option

Throws error if can't find container and can't pull it:
  - use easiest setting option

Test can launch job w/ podman and docker.
  - need to investigate zuul capability (UPDATE: request to add docker / podman to zuul [here](https://github.com/ansible/ansible-runner/issues/480))

Given a container that does exist and is pullable, but does not provide expected interface for runner to run a job in (hello-world container for example), what happens? Do we have error handling?

**Comment by jladdjr** [comment]
# in-PR Bug tracking
----------------------------
(originally from https://github.com/ansible/ansible-runner/pull/446#issuecomment-649808220)

- [ ] seems like container is failing to launch when I set `container_volume_mounts:/tmp` but does not throw error and not sure whats up -- this gets back to erroring if can't launch w/ the container runtime specified and `containerized: true`
- [ ] when job DOES launch in container, seems to not respect to correct private data dir, at least in a poetry dev env. If I have in settings file `containerized: true` and `container_runtime: docker` it DOES launch in the container, so must have found the settings file in the correct data dir, but then once inside container, is picking up old version of `demo` private data dir, so I'm thinking it is getting it from the copy of the `demo` dir in the poetry env????
- [x] not 100% if this is a "bug" but it is inconvenient that some files created by runner when running inside a container end up being owned by root. This may have something to do w/ how I have docker/podman setup so would like @jladdjr see if this happens as well (edit by @kdelee ). Files include the`$PRIVATE_DATA_DIR/.ansible` folder and `$PRIVATE_DATA_DIR/artifacts/<hash>/job_events/<hash>-partial.json`

File: ansible_runner/runner_config.py
Comment (AlanCoding) [inline ansible_runner/runner_config.py]: Is it planned to add this to the CLI? I would like to get more documentation and schema around the options. See the ansible-builder PR https://github.com/ansible/ansible-builder/pull/16

That adds:

```
$ ansible-builder build --help
usage: ansible-builder build [-h] [-t TAG] [-f FILENAME] [-b BASE_IMAGE]
                             [-c BUILD_CONTEXT]
                             [--container-runtime CONTAINER_RUNTIME]

optional arguments:
  -h, --help            show this help message and exit
  -t TAG, --tag TAG     The name for the container being built.
  -f FILENAME, --file FILENAME
                        The definiton of the execution environment.
  -b BASE_IMAGE, --base-image BASE_IMAGE
                        The parent image for the execution environment.
  -c BUILD_CONTEXT, --context BUILD_CONTEXT
                        The directory to use for the build context. Defaults
                        to $PWD/context.
  --container-runtime CONTAINER_RUNTIME
                        Specifies which container runtime to use.
```

Where the use looks like:

```
$ ansible-builder build -c alan/bc/ -f alan/alan_ee.yml --container-runtime=docker
```

I would like to see some example commands for runner, and also start syncing the schema of these options between the two projects so that they are the same interface.

I could see the `--tag` option coming into play here.

File: Dockerfile
Comment (shanemcd) [inline Dockerfile]: @AlanCoding Didnt you have some thoughts about putting the callback plugin into a collection? If we did that, then we wouldn't need this, right?

Comment (AlanCoding) [reply Dockerfile]: yes, I would absolutely like to do that

I went over what I wanted to do with Shane, and didn't see any of this container-isolation stuff as something which would fundamentally change a collection conversion.

This line, however, is something that I don't think we need. Down below, I see that `ansible-runner` gets called. If that's the entry point, then we shouldn't be setting anything like this beforehand.

That gets set in the runner code itself, only for the `ansible-playbook` process

https://github.com/ansible/ansible-runner/blob/f73705bf9b09424c9e5353d7f5119147e844eaa5/ansible_runner/runner_config.py#L198

you don't need to have it set in the process where `ansible-runner` is invoked. So just delete these 2 lines.

**Comment by AlanCoding** [comment]
github doesn't seem to automatically link this, so dropping link https://github.com/jladdjr/ansible-runner/pull/1

**Comment by jladdjr** [comment]
recheck

File: Dockerfile
Comment (AlanCoding) [inline Dockerfile:13]: @jladdjr I have https://github.com/ansible/ansible-runner/pull/473 open

I'm only committed to working in conjunction with your branch, not with devel, and I don't urgently need that change.

So I can close that if you get the change in here.

File: test/integration/containerized/priv_data/project/test-container.yml
Comment (kdelee) [inline test/integration/containerized/priv_data/project/test-container.yml]: `Look for container....`

File: test/integration/containerized/priv_data/project/test-container.yml
Comment (kdelee) [inline test/integration/containerized/priv_data/project/test-container.yml]: `Look for container....`

File: test/integration/containerized/test_cli_containerized.py
Comment (kdelee) [inline test/integration/containerized/test_cli_containerized.py:20]: :bulb: I want to add a test that runs a playbook that fails and confirms we get non-zero exit code as I could see running via the container maybe messing w/ our exit code what do you think

**Comment by jladdjr** [comment]
recheck

**Comment by jladdjr** [comment]
@kdelee :

**Comment by jladdjr** [comment]
> If user does not set container_runtime try to use first podmand then docker if neither launch, fail gracefully

@kdelee - wondering if we really want this behavior - here's an example of how I think this could lead to unexpected behavior:

1. User has been using to docker
2. .. but switches over to podman
3. When configuring ansible-runner, they specify podman, as this is now their default setup.
4. They misconfigured podman (e.g. have their registries not quite setup right, thought they had built an image, but it's not around, etc.)
5. .. but when runner switches to docker things manage to work.

If I were this user, I think I'd actually prefer things to halt with podman, since that's what I asked for, instead of it switching to another tool, that I may consider to be deprecated.

If we _do_ want this fallback behavior, think we should make the setting `default_container_runtime` to emphasize that the other know runtime will be tried.

@matburt @shanemcd - ^ let me know if you favor one scenario over the other or have any other ideas for how we should handle this

File: test/integration/containerized/test_cli_containerized.py
Comment (jladdjr) [reply test/integration/containerized/test_cli_containerized.py:20]: @kdelee - good question - I poked around some existing tests and found [this one](https://github.com/jladdjr/ansible-runner/blob/containerized_playbook_runs/test/integration/test_events.py#L14) that already asserts that the run was successful. It also does a sanity check on several events that we expect to see. What do you think about parameterizing that to run w/ and w/out containerization?

I think some of the other tests in that module may also be good candidates for parameterization if we think that running inside a container puts those other aspects of a run at risk. The `test_basic_events` test may give us the most bang for our buck, though.

Comment (jladdjr) [reply test/integration/containerized/test_cli_containerized.py:20]: got `test_basic_events` to work w/ containerization - actually helped me work out some kinks / understand how some things were connected:
https://github.com/jladdjr/ansible-runner/blob/containerized_playbook_runs/test/integration/test_events.py#L14-L55

File: .gitignore
Comment (matburt) [inline .gitignore]: poetry.lock needs to be checked in, this is kind-of the point of poetry... to make sure everyone uses the same level of dependencies.

File: Dockerfile
Comment (matburt) [reply Dockerfile]: If we do that then lets discuss it outside of the context of this PR since it's a pretty major component for Runner itself.

File: ansible_runner/__main__.py
Comment (matburt) [inline ansible_runner/__main__.py]: I can see the purpose for volume mounts here but it may be necessary to support arbitrary container parameterization (such as resource usage, network attachment, etc). Either in the form of catchall parameterization or config file.

File: ansible_runner/runner_config.py
Comment (matburt) [inline ansible_runner/runner_config.py:189]: What precipitates this inversion?

File: ansible_runner/runner_config.py
Comment (matburt) [inline ansible_runner/runner_config.py:440]: This hard-coding doesn't bring me joy 😆 

We should be able to discover the intended base. I'd like to not assume that it would *always* be `/runner` when running containerized.

File: setup.cfg
Comment (matburt) [inline setup.cfg]: While we're at it, lets just remove some of these old AWX things that stayed in when I copied this file over.

File: ansible_runner/interface.py
Comment (matburt) [inline ansible_runner/interface.py]: I wonder if "containerization" is actually the right thing to use here. This would seem to be a modification of the process isolation system and doing it this way doesn't make bwrap and container usage mutually exclusive... which it probably should be.

File: ansible_runner/runner_config.py
Comment (maxamillion) [reply ansible_runner/runner_config.py]: @AlanCoding to end game is for `ansible-builder` and `ansible-runner` to match the CLI interface for this, I have a PR that's waiting for this to merge that will "settle up" on that once I rebase.

File: ansible_runner/__main__.py
Comment (maxamillion) [reply ansible_runner/__main__.py]: +1 to that, I think trying to wrap each param we want to target for various isolation purposes will become painful to maintain quickly

File: ansible_runner/interface.py
Comment (jladdjr) [reply ansible_runner/interface.py]: @matburt That brings up a question that I've had. How much longer will we need to support both container and non-container based execution? Is that something that I could go ahead and drop in this PR?

Comment (jladdjr) [reply ansible_runner/interface.py]: If we're not going to drop non-container execution, then:

* do we want to make 'containerized' a `process_isolation` / `process_isolation_executable` setting? While containerization is a process isolation mechanism, this seems a little unintuitive.
* or, do we just want to add some logic to disable bwrap when `containerized: True`? I'm guessing this is the middle-ground that we'll end up going with.

File: ansible_runner/runner_config.py
Comment (jladdjr) [reply ansible_runner/runner_config.py:440]: @matburt - I was looking at this yesterday. We're mapping the private data dir on the host to `/runner` in the container, which is what allows us to hardcode `/runner` as the base. That _seems_ like it should be okay, but maybe that's not actually functionally equivalent. Let me know if you would still consider `/runner` to be hard mapping given the volume mapping.

File: ansible_runner/runner_config.py
Comment (shanemcd) [reply ansible_runner/runner_config.py:189]: This one was from me... 😄 Basically, `self.inventory` gets set inside of `prepare_inventory` and I kinda needed that to be set (properly) before building the command.

File: ansible_runner/interface.py
Comment (shanemcd) [reply ansible_runner/interface.py]: I am definitely not a fan of changing this option to `containerization`. `ansible-runner run --containerized` makes way more sense than `ansible-runner run --containerization`. Am I missing something?

And I think we should just throw an error if you try to use `--process-isolation` at the same time as `--containerized`.

File: ansible_runner/runner_config.py
Comment (matburt) [reply ansible_runner/runner_config.py:440]: This *is* the private data dir, though and it should be set/fetched from that:

https://github.com/ansible/ansible-runner/blob/devel/ansible_runner/runner_config.py#L88

Comment (matburt) [reply ansible_runner/runner_config.py:440]: Actually, I retract that... because I see the attachment locality of this now. I'm still not sure how I feel about this in the general sense and it might should be based on where we know we are attaching volumes if the person is doing something *other* than what we expected them to do.

**Comment by jladdjr** [comment]
* Refactored the process isolation options to fold in containerization as a possible method.
* Added `container_options` as a catch-all for any other options the user might want to pass to the container engine.

File: Dockerfile
Comment (AlanCoding) [reply Dockerfile]: Clarification: we should remove these settings, because they are set in the code itself.

File: test/integration/containerized/priv_data/env/settings
Comment (AlanCoding) [inline test/integration/containerized/priv_data/env/settings]: @kdelee this look wrong, it was changed but this isn't the syntax it changed to. Worked for me:

```yaml
---
container_image: ansible-execution-env
process_isolation_executable: docker
process_isolation: true
```

https://github.com/ansible/ansible-builder/pull/43

Comment (jladdjr) [reply test/integration/containerized/priv_data/env/settings]: > @kdelee this look wrong, it was changed but this isn't the syntax it changed to. Worked for me:

I'm having trouble parsing this. The example you posted does reflect the latest options, though.

File: ansible_runner/interface.py
Comment (jladdjr) [reply ansible_runner/interface.py]: @shanemcd and I caught up about this. We ended up with these options, which folds containerization into the existing process isolation options, and adds a few options specific to containerization:

```
--process-isolation => {true, false}
--process-isolation-executable => {bwrap, podman, docker}
--container-image => defaults to `ansible/ansible-runner`
--container-volume-mounts => list of host:container dir mappings
--container-options => list of any other options you may want to pass to the execution engine
```

File: ansible_runner/__main__.py
Comment (jladdjr) [reply ansible_runner/__main__.py]: added `--container-options` as a catch-all for any remaining options the user may want to pass to runner

**Comment by jladdjr** [comment]
@kdelee - dropping the 'create env var' todo's from our list. I thought that env vars were something that runner exposed for most options. What I was (mis)remembering though was that when I did the [perf stats work](https://github.com/ansible/ansible-runner/pull/346/files#diff-c62b92cfe059167de0305220ea16db97R218), I used env vars to talk with the [cgroup_perf_recap plugin](https://docs.ansible.com/ansible/latest/plugins/callback/cgroup_perf_recap.html).

**Comment by jladdjr** [comment]
@shanemcd, @matburt - one of the items that @kdelee and I thought of adding was a flag that would instruct runner not to pull down an image if one wasn't already available. Do you think this would be useful?

If so, I'm not seeing an obvious option for `docker run` that would block pulling an image. I suppose what we could do instead is check `docker images` to see if the image is already available, and stop if one isn't.

**Comment by jladdjr** [comment]
> if container does not have whatever we need to call ansible-playbook, handle error gracefully and place something in the artifacts file and not just error to stdout so that when user that consumes artifacts to find results, they have evidence that helps them find out what happened.

@kdelee - Confirmed that runner provides helpful artifacts in the event that the container is missing deps. Here are the artifacts left behind when ansible itself is missing from the container. All of these are located under `artifacts/<uuid>`.

---

**job_events/1-f240c47b-aa7e-4153-8df2-d83a77df7988.json**
```
{
  "event": "verbose",
  "uuid": "f240c47b-aa7e-4153-8df2-d83a77df7988",
  "counter": 1,
  "stdout": "[FATAL tini (12)] exec ansible-playbook failed: No such file or directory",
  "start_line": 0,
  "end_line": 1,
  "runner_ident": "05725d1d-3149-40eb-a174-11ee1b2c3310"
}
```

**rc**
`127`

**status**
`failed`

**stdout**
`[FATAL tini (12)] exec ansible-playbook failed: No such file or directory`

File: test/integration/test_events.py
Comment (chrismeyersfsu) [inline test/integration/test_events.py:32]: Using `**` vs. explicitly listing the key/value params will build up technical debt. We had to pay this dept in AWX when we added runner support to `tasks.py`

**Comment by jladdjr** [comment]
Looking for any final feedback before merge. I _think_ that everything has been covered (apart from adding some more docs, which I'm working on right now), but if I've missed anything let me know!

**Comment by AlanCoding** [comment]
Before we merge this, I think we should figure out what the cause of the warning is:

```
[DEPRECATION WARNING]: 'awx_display' is subclassing DefaultCallback without the
 corresponding doc_fragment. This feature will be removed from ansible-base in 
version 2.14. Deprecation warnings can be disabled by setting 
deprecation_warnings=False in ansible.cfg.
```

That is specific to this PR, and there's a logical leap from what it changed to getting the wrong callback somehow.

**Comment by AlanCoding** [comment]
I re-ran one of my examples with the playbook:

```yaml
---
- hosts: localhost
  gather_facts: false
  connection: local
  tasks:
    - debug: msg="hello"

    - command:
        cmd: printenv
```

This finds the value

```
ANSIBLE_CALLBACK_PLUGINS=/Users/alancoding/Documents/repos/ansible-runner/ansible_runner/callbacks
```

That pretty well confirms the theory, that even though the container sets `ENV` (which it shouldn't), it doesn't matter because the ansible-runner parent process will overwrite that before it calls pexpect to create the subprocess. When it does that, it uses a location on _my_ computer, which obviously doesn't exist in the container.

Even if we get that, I don't see how we run the current ansible-runner tests containerized.

**Comment by AlanCoding** [comment]
Put up a proposed fix:

https://github.com/jladdjr/ansible-runner/pull/3

But again, I think we really need to ask the question of how we test against regressions like this?

File: ansible_runner/runner_config.py
Comment (AlanCoding) [inline ansible_runner/runner_config.py]: What about `container_options`? shouldn't you debug those here too?

Comment (maxamillion) [reply ansible_runner/runner_config.py]: +1 - container options should probably be debugged here, that and/or the full container invocation command. Maybe log `new_args` unless there's a concern of logging/leaking secret data.

**Comment by AlanCoding** [comment]
I'm now thinking that the subclassing error message may be https://github.com/ansible-collections/community.general/issues/565, which was fixed on the Ansible core side, and the version in the base image is just out of date.

I also confirmed that the fix https://github.com/ansible/ansible/pull/70228 was not applied to the source that I inspected inside of the container.

File: ansible_runner/__main__.py
Comment (shanemcd) [inline ansible_runner/__main__.py]:
```suggestion
        default="podman",
```

Comment (jladdjr) [reply ansible_runner/__main__.py]: sounds like some kind of super-hero - fixed

**Review by maxamillion — APPROVED** [review]

**Comment by jladdjr** [comment]
> I'm now thinking that the subclassing error message may be ansible-collections/community.general#565, which was fixed on the Ansible core side, and the version in the base image is just out of date.

@AlanCoding - I just built a container that installed ansible-base from source and confirmed I didn't see the warning.

File: ansible_runner/runner_config.py
Comment (jladdjr) [reply ansible_runner/runner_config.py]: Good catch - updated

File: ansible_runner/__main__.py
Comment (ryanpetrello) [inline ansible_runner/__main__.py:507]: `append` I like this API, nice 👍

File: ansible_runner/runner_config.py
Comment (ryanpetrello) [inline ansible_runner/runner_config.py:208]: Ignoring the AWX use case, and considering people who might just install and use runner's CLI, this assumes you've installed runner in your global `site-packages`, which might break some runner use cases (what about people who install runner into a virtualenv?)

File: test/integration/conftest.py
Comment (ryanpetrello) [inline test/integration/conftest.py:23]: Yes, we should totally do this 😄

**Review by ryanpetrello — APPROVED** [review]
minimal feedback, 👍 this is pretty darn fancy 😄

File: ansible_runner/runner_config.py
Comment (AlanCoding) [reply ansible_runner/runner_config.py:208]: That gets messy because venv paths become part of the software interface. Job isolation will not be compatible with all images that have Ansible, we thought that EEs would have a fairly minimal set of requirements (ansible / runner), but I'm questioning that as we get into the details.

If you're running an image that uses venvs, then how does that get communicated from the image to `ansible-runner`? Even if the image itself had some metadata associated with it, runner would have to run docker/podman commands to find that out. See https://github.com/ansible/ansible-builder/issues/33 about venvs. 

This is a potential solution, take it with a _large_ grain of salt:

 - In the base image, copy the runner callback stuff to `/usr/share/ansible/plugins/callback`, since this is a default value of `ANSIBLE_CALLBACK_PLUGINS`
 - if the job is containerized, don't set this path at all, because there's no need to.

Considering the import hacks in the callback, I anticipate that this may take some work to get right. On the other hand... if we put the callback into a collection inside of `/usr/share/collections` this would all start to make a lot more sense.

**Review by shanemcd — APPROVED** [review]

**Comment by jladdjr** [comment]
recheck

**Comment by jladdjr** [comment]
recheck

File: ansible_runner/runner_config.py
Comment (shanemcd) [inline ansible_runner/runner_config.py]:
```suggestion
        if self.containerized:
```

**Comment by jladdjr** [comment]
recheck

**Pull request merged**
